atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TableSizeCache.java

```diff
@@ -15,41 +15,74 @@
  */
 package com.palantir.atlasdb.keyvalue.dbkvs.impl;
 
+import java.sql.SQLException;
 import java.util.concurrent.ExecutionException;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.google.common.base.Preconditions;
-import com.google.common.cache.Cache;
+import com.google.common.base.Throwables;
 import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Iterables;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
-import com.palantir.common.base.Throwables;
 import com.palantir.nexus.db.sql.AgnosticResultSet;
+import com.palantir.nexus.db.sql.SqlConnection;
 
 public final class TableSizeCache {
-    private static final Cache<TableReference, TableSize> tableSizeByTableRef = CacheBuilder.newBuilder().build();
+    private static final Logger log = LoggerFactory.getLogger(TableSizeCache.class);
+
+    private final ConnectionSupplier conns;
+    private final TableReference metadataTable;
 
-    private TableSizeCache() {
-        //Utility class
+    public TableSizeCache(ConnectionSupplier conns, TableReference metadataTable) {
+        this.conns = conns;
+        this.metadataTable = metadataTable;
     }
 
-    public static TableSize getTableSize(
-            final ConnectionSupplier conns,
-            final TableReference tableRef,
-            TableReference metadataTable) {
+    private LoadingCache<TableReference, TableSize> tableSizeByTableRef = CacheBuilder.newBuilder().build(
```

Comment:
yeah, there was a large problem here that the cache would get created each time but only used once so had to return it to the static methodology it had before.